File: QA-Assignment/BusinessWorkflowAutomation/pages/dashboard_page.py
import re
from playwright.sync_api import Page, expect
from pages.base_page import BasePage


class DashboardPage(BasePage):
    """Page Object Model for the Dashboard Page (/dashboard)."""

    def __init__(self, page: Page):
        super().__init__(page)
        self.welcome_heading = page.get_by_role("heading", name=re.compile(r"Welcome,", re.IGNORECASE))
        self.new_chat_button = page.get_by_test_id("new-chat-button")
        self.sidebar_logout_button = page.get_by_test_id("sidebar-logout-button")

    def goto(self):
        """Navigate to Dashboard Page directly."""
        self.navigate_to("/dashboard")
        self.assert_on_dashboard()

    def assert_on_dashboard(self):
        """Assert Dashboard page is active and welcome banner is visible.
        
        If user landed on a workspace page after signup, navigate back to /dashboard.
        Note: page.url is a property (not a method) in Python Playwright.
        """
        current_url = self.page.url
        if "/dashboard" not in current_url:
            self.navigate_to("/dashboard")
        expect(self.page).to_have_url(re.compile(r".*/dashboard"), timeout=15000)
        expect(self.welcome_heading).to_be_visible(timeout=15000)

    def click_new_chat(self):
        """Click New Chat button and verify navigation to chat workspace."""
        current_url = self.page.url
        if "/dashboard" not in current_url:
            self.navigate_to("/dashboard")
            expect(self.page).to_have_url(re.compile(r".*/dashboard"), timeout=15000)
        expect(self.new_chat_button).to_be_visible(timeout=10000)
        self.new_chat_button.click()
        expect(self.page).to_have_url(re.compile(r".*/workspace/chat/.*"), timeout=15000)

    def click_add_website(self):
        """Navigate directly to Knowledge Base indexing page."""
        self.navigate_to("/workspace/index")
        expect(self.page).to_have_url(re.compile(r".*/workspace/index"), timeout=15000)

    def logout(self):
        """Perform user logout from any page (Dashboard or Workspace)."""
        logout_by_text = self.page.locator("button", has_text=re.compile(r"log.?out", re.IGNORECASE))
        
        if self.sidebar_logout_button.count() > 0 and self.sidebar_logout_button.first.is_visible():
            self.sidebar_logout_button.first.dispatch_event("click")
        elif logout_by_text.count() > 0 and logout_by_text.first.is_visible():
            logout_by_text.first.dispatch_event("click")
        else:
            profile_trigger = self.page.locator('button[title="Account profile"]').or_(
                self.page.locator("button", has_text=re.compile(r"QA Tester|@gmail\.com", re.IGNORECASE))
            )
            if profile_trigger.count() > 0 and profile_trigger.first.is_visible():
                profile_trigger.first.click()
                self.page.wait_for_timeout(500)
            
            logout_btn = self.page.get_by_test_id("logout-button").or_(logout_by_text)
            expect(logout_btn.first).to_be_visible(timeout=8000)
            logout_btn.first.dispatch_event("click")

        expect(self.page).to_have_url(re.compile(r".*/login"), timeout=15000)

File: QA-Assignment/Automation/pages/DashboardPage.js
import { expect } from '@playwright/test';

/**
 * Page Object Model for the Dashboard Page (/dashboard)
 */
export class DashboardPage {
  /**
   * @param {import('@playwright/test').Page} page
   */
  constructor(page) {
    this.page = page;

    // Locators matching Dashboard UI
    this.welcomeHeading = page.getByRole('heading', { name: /Welcome,/i });
    // New chat button – exists both on /dashboard AND in the sidebar on workspace pages
    this.newChatButton = page.getByTestId('new-chat-button');
    // Sidebar new-chat button (button text "New chat" in sidebar)
    this.sidebarNewChatButton = page.locator('button', { hasText: /^New chat$/i });
    this.recentChatsHeading = page.getByRole('heading', { name: 'Recent Chats' });
    this.knowledgeBasesHeading = page.getByRole('heading', { name: 'Knowledge Bases' });
    // Add website – matches both "+ Add Website" on dashboard and "Add website" in sidebar
    this.addWebsiteLink = page.locator('button', { hasText: /add website/i }).first();
    // Profile dropdown: confirmed by live inspection — uses title="Account profile"
    this.profileDropdownTrigger = page.locator('button[title="Account profile"]');
    // Logout: sidebar button has confirmed data-testid="sidebar-logout-button"
    this.logoutButton = page.getByTestId('sidebar-logout-button');
  }

  /**
   * Navigate directly to Dashboard page
   */
  async goto() {
    await this.page.goto('/dashboard');
    await this.assertOnDashboard();
  }

  /**
   * Assert Dashboard page is loaded.
   * Works from dashboard URL or any workspace URL (sidebar is always present after login).
   */
  async assertOnDashboard() {
    // Accept either /dashboard or /workspace/* — both mean user is authenticated
    await expect(this.page).toHaveURL(/\/(dashboard|workspace)/, { timeout: 15000 });
    // The welcome heading only appears on /dashboard — navigate there if needed
    const currentUrl = this.page.url();
    if (!currentUrl.includes('/dashboard')) {
      await this.page.goto('/dashboard');
      await expect(this.page).toHaveURL(/.*\/dashboard/, { timeout: 15000 });
    }
    await expect(this.welcomeHeading).toBeVisible({ timeout: 15000 });
  }

  /**
   * Click New Chat button — navigates to /workspace/chat/:id
   */
  async clickNewChat() {
    // Ensure we are on /dashboard before clicking the big "New chat" button
    const currentUrl = this.page.url();
    if (!currentUrl.includes('/dashboard')) {
      await this.page.goto('/dashboard');
      await expect(this.page).toHaveURL(/.*\/dashboard/, { timeout: 15000 });
    }
    await expect(this.newChatButton).toBeVisible({ timeout: 10000 });
    await this.newChatButton.click();
    await expect(this.page).toHaveURL(/.*\/workspace\/chat\/.*/, { timeout: 15000 });
  }

  /**
   * Click Add Website link to navigate to indexing page
   */
  async clickAddWebsite() {
    // Navigate directly to /workspace/index — avoids button click race condition
    await this.page.goto('/workspace/index');
    await expect(this.page).toHaveURL(/.*\/workspace\/index/, { timeout: 15000 });
  }

  /**
   * Perform logout via Profile Dropdown (title="Account profile" or Log out button in sidebar)
   */
  async logout() {
    // The sidebar shows a direct "Log out" button — try that first
    const directLogout = this.page.locator('button', { hasText: /log.?out/i });
    if (await directLogout.count() > 0 && await directLogout.first().isVisible()) {
      await directLogout.first().click();
    } else {
      // Fallback: profile dropdown trigger
      await this.profileDropdownTrigger.click();
      await directLogout.waitFor({ state: 'visible', timeout: 8000 });
      await directLogout.first().click();
    }
    await expect(this.page).toHaveURL(/.*\/login/, { timeout: 15000 });
  }
}
